File: go/quicksort1/quicksort1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	//"time"
	"flag"
)

var (
	memprofile = flag.String("memprofile", "", "write memory profile to this file")
	cpuprofile = flag.String("cpuprofile", "", "write cpu profile to this file")
	infile     = flag.String("infile", "../data/array/array.in", "Matrices to multiply")
)

func partition(arr []int64, low, high int) int {
	pivot := arr[high]
	i := low - 1

	for j := low; j < high; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

func quickSort(arr []int64, low, high int) {
	if low < high {
		pivot := partition(arr, low, high)
		quickSort(arr, low, pivot-1)
		quickSort(arr, pivot+1, high)
	}
}

func main() {
	flag.Parse()
	
	file, _ := os.Open(*infile)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	scanner.Scan()
	arr_length := strings.Fields(scanner.Text())

	n, _ := strconv.Atoi(arr_length[0])

	arr := make([]int64, n)
	scanner.Scan()
	line := scanner.Text()
	iss := strings.Fields(line)
	
	for i, numStr := range iss{
		arr[i], _ = strconv.ParseInt(numStr, 10, 64)
	}

	//start := time.Now()
	quickSort(arr, 0, n-1)
	//end := time.Now()

	for _, element := range arr {
		fmt.Println(element)
	}

	//fmt.Println("Mergesort time for", n, "elements:", end.Sub(start))
}

File: cpp/quicksort1/quicksort1.cc
#include <bits/stdc++.h>
#include <gperftools/profiler.h>
#include <ctime>
using namespace std;

int partition(std::vector<long long>& arr, int low, int high)
{
    int pivot = arr[high];
    int i = (low-1);

    for(int j=low; j<=high; j++){
        if(arr[j] < pivot){
            i++;
            swap(arr[i],arr[j]);
        }
    }
    swap(arr[i+1], arr[high]);
    return (i+1);
}

void quickSort(std::vector<long long>& arr, int low, int high){
    if(low < high){
        int pivot = partition(arr,low,high);
        quickSort(arr,low,pivot-1);
        quickSort(arr,pivot+1,high);
    }
}

int main(){
    fstream f("../data/array/array.in");
    if (!f.good()){
        std::cerr << "double check file path\n";
        return 1;
    }

    string line;
    getline(f,line);
    istringstream arr_length(line);
    int n;
    arr_length >> n;
    //cout << n << endl;

    long long num;
    getline(f,line);
    std::istringstream iss(line);
    std::vector<long long> arr;
    while (iss >> num){
        arr.push_back(num);
    }
    
    clock_t start = clock();
    quickSort(arr, 0, n-1);
    clock_t end = clock();
    
    for (auto &element : arr){
        cout << element << endl;
    }

    //cout<< "Quicksort time for " << n << " elements: "<< (float)(end - start)/ CLOCKS_PER_SEC << " "<<endl;;
    f.close();
    return 0;
}